monai/data/image_reader.py

```diff
@@ -0,0 +1,348 @@
+                if not np.allclose(header["spatial_shape"], compatible_meta["spatial_shape"]):
+                    raise RuntimeError("spatial_shape of all images should be same.")
+
+        img_array = np.stack(img_array, axis=0) if len(img_array) > 1 else img_array[0]
+        return img_array, compatible_meta
+
+    def _get_meta_dict(self, img: itk.Image) -> Dict:
+        """
+        Get all the meta data of the image and convert to dict type.
+
+        Args:
+            img: a ITK image object loaded from a image file.
+
+        """
+        img_meta_dict = img.GetMetaDataDictionary()
+        meta_dict = dict()
+        for key in img_meta_dict.GetKeys():
+            # ignore deprecated, legacy members that cause issues
+            if key.startswith("ITK_original_"):
+                continue
+            meta_dict[key] = img_meta_dict[key]
+        meta_dict["origin"] = np.asarray(img.GetOrigin())
+        meta_dict["spacing"] = np.asarray(img.GetSpacing())
+        meta_dict["direction"] = itk.array_from_matrix(img.GetDirection())
+        return meta_dict
+
+    def _get_affine(self, img: itk.Image) -> np.ndarray:
+        """
+        Get or construct the affine matrix of the image, it can be used to correct
+        spacing, orientation or execute spatial transforms.
+        Construct Affine matrix based on direction, spacing, origin information.
+        Refer to: https://github.com/RSIP-Vision/medio
+
+        Args:
+            img: a ITK image object loaded from a image file.
+
+        """
+        direction = itk.array_from_matrix(img.GetDirection())
+        spacing = np.asarray(img.GetSpacing())
+        origin = np.asarray(img.GetOrigin())
+
+        direction = np.asarray(direction)
+        affine = np.eye(direction.shape[0] + 1)
+        affine[(slice(-1), slice(-1))] = direction @ np.diag(spacing)
+        affine[(slice(-1), -1)] = origin
+        return affine
+
+    def _get_spatial_shape(self, img: itk.Image) -> Sequence:
+        """
+        Get the spatial shape of image data, it doesn't contain the channel dim.
+
+        Args:
+            img: a ITK image object loaded from a image file.
+
+        """
+        return list(itk.size(img))
+
+    def _get_array_data(self, img: itk.Image) -> np.ndarray:
+        """
+        Get the raw array data of the image, converted to Numpy array.
+
+        Args:
+            img: a ITK image object loaded from a image file.
+
+        """
+        return itk.array_view_from_image(img, keep_axes=self.keep_axes)
+
+
+class NibabelReader(ImageReader):
+    """
+    Load NIfTI format images based on Nibabel library.
+
+    Args:
+        img: image to initialize the reader, this is for the usage that the image data
+            is already in memory and no need to read from file again, default is None.
+        as_closest_canonical: if True, load the image as closest to canonical axis format.
```

Comment:
Should the APIs for these readers be as identical as possible?   If ITKReader has a keep_axis/use_c_order_axis/ whatever, then should the other readers in the main MONAI system have those same arguments whenever possible?   I realize this may be awkward and have a computational cost for some readers to support, but particularly for axis ordering, having this argument in every reader would avoid nasty surprises if a different readers use different orderings, etc.

Reply:
Hi @aylward and @wyli ,

I am not sure whether this `keep_axes` and `as_closest_canonical` are specific args for ITK and Nibabel?
If you guys think they should be unified for all the readers, I can change to `ImageReader`, what do you think?

Thanks.

Reply:
In other interfaces providing common access to various components the function or constructor would keyword arguments (called `**other_args` or something similar) to be passed along to the actual internal object. Matplotlib does this in places but relies on documentation to state which values are valid.

Reply:
keep_axes and as_closest_canonical are all about particular 'views' of the underlying storage,  so shall we have
```def get_data(self, **kwargs)``` in the `ImageReader`,
 and in the specific reader implementation we could have various args such as
```py
def get_data(self, keep_axes=, dtype=, **other_args)
def get_data(self, as_closest_canonical=, dtype=, **other_args)
def get_data(self, coordinates=, dtype=, **other_args) # lazy access to some partial images by coordinates?
```

Reply:
Hi @wyli ,

In your proposal, what's the `**other_args` will be used in a particular reader?
I think @ericspod means we use `def get_data(self, **kwargs)` for both `ImageReader` and particular readers.
And document what is valid args to pass in every reader.
What do you think?

Thanks.

Reply:
Alternatively replace the constructor arguments which aren't common to all subtypes of `ImageReader` with `**other_args` such as `keep_axes` or `as_closest_canonical`. The idea is to have a common interface everywhere, whether that's in the constructor or methods like `get_data`.

Reply:
I meant to use **other_args to tolerate the irrelevant arguments and the named arguments make the method slightly easier to read. but I also agree with @ericspod's solution. btw do we really need an ``__init__`` in `class ImageReader`?

Reply:
I updated in the latest commit according to @ericspod 's suggestion.
I think at least we can keep `__init__` in `ImageReader` to do overall logic for all readers later, just like our `Transform`.
Thanks.

Reply:
currently `Transform` https://github.com/Project-MONAI/MONAI/blob/master/monai/transforms/compose.py#L26 doesn't have `__init__`

Reply:
OK, so you suggest to remove `__init__` in `ImageReader` and use `keep_axes` or `as_closest_canonical` as args directly for `ITKReader` and `NibabelReader`, no need `other_args` anymore?

Reply:
After review the code, I removed `__init__` in `ImageReader` and uses `keep_axes` and `as_closest_canonical` in readers directly according latest discussion (which may be against @aylward 's original comment).
I think it's easy to be compatible with user's program even we changed to `__init__(img, **other_args)` in future because user still call it by `ITKReader(img, keep_axes=True)`.

Thanks.

Reply:
based on the previous discussions I'd implement the reader as:
```py
class ITKReader(ImageReader):
    self._img = None

    def read(self, data: Any, **kwargs):
        # kwargs may provide additional args to itk.imread
        if data is string-like filename:
            self._img = self.read_by_file_name(data)
        if data is string like directory name:
            self._img = self.read_by_dir_name(data)
        if data is itk img object:
            self._img = data

    def get_data(self, dtype=float, **kwargs):
        # 1. if self._img is None: please call self.read first
        # 2. get array data from self._img with kwargs
        # 3. return data, meta_data
```
any concerns about this?

Reply:
Hi @wyli ,

Thanks for your suggestion, I think we haven't discussed `read dir` behavior so far, can we add it later?
Others look good to me.

Thanks.

Reply:
sure thanks @Nic-Ma